File: DataStructure/JavaSolution/Chapter 3 Stack&Queue/Queue/Queue/src/Queue/ArrayQueue.java
package Queue;

public class ArrayQueue<E> implements IQueue<E> {
    
    private Array<E> array;

    public ArrayQueue(int capacity) {
        array = new Array<E>(capacity);
    }

    public ArrayQueue() {
        this(20);
    }
    
    public int getSize() {
        return array.getSize();
    }
    public void enqueue(E e) {
        array.addLast(e);
    }
    public E dequeue() {
       return array.removeFirst();
    }
    public boolean isEmpty() {
        return array.isEmpty();
    }
    public E getFront() {
        return array.getFirst();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(String.format("Size: %d, Capacity: %d\n", array.getSize(), array.getCapacity()));
        sb.append("Top [");
        for (int i = 0; i < array.getSize(); i++) {
            sb.append(array.get(i));
            if (i != array.getSize() - 1) {
                sb.append(", ");
            }
        }
        sb.append("] Tail");

        return sb.toString();
    }
}

File: MoshDataStructure&Algrithm/DataStructureAndAlgorithmMosh/ArrayQueue.cs
using System;
using System.Text;

namespace QueueDemo {
    public class ArrayQueue {
        private int[] items;
        private int front;
        private int rear;
        private int count;

        public ArrayQueue (int capacity) {
            if (capacity <= 0) throw new ArgumentOutOfRangeException ("Invalid Capacity");

            items = new int[capacity];
            front = 0;
            rear = 0;
            count = 0;
        }

        public void Enqueue (int item) {
            if (IsFull ()) throw new Exception ("Queue is full");
            items[rear++] = item;
            count++;
        }

        public int Dequeue () {
            if (IsEmpty ()) throw new Exception ("Queue is Empty");
            int result = items[front];
            items[front++] = 0;
            count--;
            return result;
        }

        public bool IsFull () {
            return count == items.Length;
        }

        public bool IsEmpty () {
            return count == 0;
        }

        public override string ToString () {
            StringBuilder sb = new StringBuilder ();
            sb.Append ("[");
            for (int i = 0; i < items.Length; i++) {
                sb.Append ($"{items[i]} ");
            }
            sb.Append ("]");
            return sb.ToString ();
        }
    }

    public class ArrayCircularQueue {
        private int[] items;
        private int front;
        private int rear;
        private int count;

        public ArrayCircularQueue (int capacity) {
            if (capacity <= 0) throw new ArgumentOutOfRangeException ("Invalid Capacity");

            items = new int[capacity];
            front = 0;
            rear = 0;
            count = 0;
        }

        public void Enqueue (int item) {
            if (IsFull ()) throw new Exception ("Queue is full");
            items[rear] = item;
            rear = (rear + 1) % items.Length;
            count++;
        }

        public int Dequeue () {
            if (IsEmpty ()) throw new Exception ("Queue is Empty");
            int result = items[front];
            items[front] = 0;
            front = (front + 1) % items.Length;
            count--;
            return result;
        }

        public bool IsFull () {
            return count == items.Length;
        }

        public bool IsEmpty () {
            return count == 0;
        }

        public override string ToString () {
            StringBuilder sb = new StringBuilder ();
            sb.Append ("[");
            for (int i = 0; i < items.Length; i++) {
                sb.Append ($"{items[i]} ");
            }
            sb.Append ("]");
            return sb.ToString ();
        }
    }
}
